man/fi_verbs.7.md

```diff
@@ -88,6 +88,20 @@ for restrictions. It can be turned off by setting the FI_FORK_UNSAFE environment
 variable to "yes". This can improve the performance of memory registrations but it
 also makes the use of fork unsafe.
 
+### Memory Registration Cache
+The verbs provider features a memory registration cache. This speeds up memory
+registration calls from applications by caching registrations of frequently used
+memory regions. It also has the option of lazy memory registration cache where
+registrations of memory regions only above a certain size get cached. The user
```

Comment:
Slight wording tweak starting at "It also has the option...":
Cached registrations are lazily unregistered, which allows registration calls of the same region to only go through the registration process once.  Additional controls allow caching only registrations of memory regions under a specified size.

Reply:
@gladkovdmitry17 From the code it seems when `FI_VERBS_MR_CACHE_LAZY_SIZE` is set, memory registrations for buffers only above a certain size would be cached:
https://github.com/ofiwg/libfabric/blob/2640bcca250548feb713cf0a7add2e5075fa5d03/prov/verbs/src/verbs_mr.c#L332

Is that the intended behavior? @shefty and I discussed about this and it seems we should cache memory registration until a certain size and not the other way around. An application might use small buffers many times than a larger one.

Reply:
Yes, that is the intended behavior. The larger the buffer, the more expensive it is to register, isn't it? From our experiments it should be so

Reply:
Registration costs are related to the number of pages being registered, but the cost of registration relative to the cost of the transfer is greater for smaller messages.  That is, I believe you'll see a bigger drop in performance doing: register 4k + send 4k + unregister 4k, than you will if the buffer size were 1 GB.

Pinning large regions will impact how much memory is available across the system, so it affects all processes.  So, for large transfers, I don't think you want to maintain pinning that memory longer than is necessary.  I also wouldn't expect an app to re-use a 1 GB buffer, for example, versus reusing a 4k buffer.

This is why I would expect to see limitations placed on how large a region can be and still be cached, as well as a possible limit on the total amount of memory cached at once.

Reply:
> Registration costs are related to the number of pages being registered,
> but the cost of registration relative to the cost of the transfer is greater for smaller messages.

Yes, I agree.

> Pinning large regions will impact how much memory is available across the system

Sure, but our goal is to reduce cost that we need to pay when sending large regions. Unfortunately, this scheme is not good, but it works.

Take into consideration our discussions., I'd say the following:
The idea of having both `FI_VERBS_MR_CACHE_LAZY_SIZE` and `FI_VERBS_MR_CACHE_SIZE` variables is  improper.
Why we should count how many entries were registered and limit size from which  provider should cache regions?  let's set  the threshold - how much memory can be cached?
For example, 1 GB. 
User can register 1 GB region and other regions won't be cached. But as an opposite example - user can register 1024 regions of 1024 MB... and etc. But there is limit - 1 GB. It will be simply controlled, we will have a variable that will contain size (how much memory is registered) and it will be limited by threshold.

 I guess, my idea is clear :) What do you think? Sorry that this idea didn't come to me before :)

Reply:
I like the idea of having a single threshold that will limit the total amount of memory registered by the cache.

Reply:
> I like the idea of having a single threshold that will limit the total amount of memory registered by the cache.

Thanks, Sean
So, I'll do that